### PR TITLE
[SPARK-585] Configurable deployment of Dispatcher on UCR

### PR DESCRIPTION
**Pull request by ArtRand**
Use the Mesos Universal Container Runtime (UCR) for the dispatcher. This allows `dcos task exec`. The health check and Spark signal handler also work better now.

### CODE REVIEW

**Comment by ArtRand** [comment]
@skonto try this when you have a second?

File: universe/marathon.json.mustache [hunk 1]
```diff
@@ -46,49 +48,92 @@
...
-{{#security.ssl.enabled}}
-            "protocol": "COMMAND",
-            "command": { "value": "curl -f -k https://$HOST:$PORT2/" },
-{{/security.ssl.enabled}}
-{{^security.ssl.enabled}}
-            "path": "/",
             "portIndex": 2,
-            "protocol": "HTTP",
-{{/security.ssl.enabled}}
+            "protocol": "MESOS_HTTP",
```
Comment (susanxhuynh) [inline universe/marathon.json.mustache]: Just curious, is this change independent of the UCR change?

File: universe/marathon.json.mustache [hunk 2]
```diff
@@ -46,49 +48,92 @@
...
     ],
+
+    "upgradeStrategy": {
+        "minimumHealthCapacity": 0,
+        "maximumOverCapacity": 0
+    },
+
+    "unreachableStrategy": {
+        "expungeAfterSeconds": 0,
+        "inactiveAfterSeconds": 0
```
Comment (susanxhuynh) [inline universe/marathon.json.mustache]: The default value is 300 sec. Why did you change it to zero?

File: universe/marathon.json.mustache (hunk 1 above)
Comment (ArtRand) [reply universe/marathon.json.mustache]: I picked up this change from @vishnu2kmohan actually. It makes the health check in the Mesos UI show up green.

File: universe/marathon.json.mustache (hunk 2 above)
Comment (ArtRand) [reply universe/marathon.json.mustache]: We store the state in ZK, so might as well restart immediately.

File: universe/marathon.json.mustache [hunk 3]
```diff
@@ -28,7 +29,8 @@
 {{#service.spark-history-server-url}}
         "SPARK_HISTORY_SERVER_URL": "{{service.spark-history-server-url}}",
 {{/service.spark-history-server-url}}
-
+        "JAVA_HOME": "/usr/lib/jvm/jre1.8.0_131",
```
[the remark this answers is not in the document]
Comment (ArtRand) [reply universe/marathon.json.mustache]: which?

[the remark this answers is not in the document]
Comment (ArtRand) [reply universe/marathon.json.mustache]: If you mean the mustache template below, it's mean "if not set"

File: universe/marathon.json.mustache (hunk 2 above)
Comment (ArtRand) [reply universe/marathon.json.mustache]: if you feel strongly I can change it back

File: universe/marathon.json.mustache [hunk 4]
```diff
@@ -46,49 +48,92 @@
...
-        0
+    "portDefinitions": [
+        {
+            "port": 0,
+            "protocol": "tcp",
+            "name": "dispatcher",
+            "labels": {
+                "VIP_0": "spark-dispatcher:7077"
+            }
+        },
```
[the remark this answers is not in the document]
Comment (ArtRand) [reply universe/marathon.json.mustache]: I don't think so, because of the way ports are handed in the CLI right now. I think it probably can be, eventually, but not in this PR.

File: universe/marathon.json.mustache (hunk 3 above)
[the remark this answers is not in the document]
Comment (susanxhuynh) [reply universe/marathon.json.mustache]: I meant the two new env vars that were added (JAVA_HOME, LD_LIBRARY_PATH). I wasn't sure why they were added.

File: universe/marathon.json.mustache (hunk 1 above)
Comment (susanxhuynh) [reply universe/marathon.json.mustache]: Ok

File: universe/marathon.json.mustache (hunk 2 above)
Comment (susanxhuynh) [reply universe/marathon.json.mustache]: Makes sense. I was mistakenly thinking this applied to the drivers instead of the dispatcher.

File: universe/marathon.json.mustache (hunk 3 above)
[the remark this answers is not in the document]
Comment (ArtRand) [reply universe/marathon.json.mustache]: Oh, sorry. `JAVA_HOME` and `LD_LIBRARY_PATH` allow you to run `spark-shell` and `spark-submit` in client mode with `dcos task exec`. For cluster mode we need to remove the `MESOS_*` environment variables, which needs to happen in Spark core.

[the remark this answers is not in the document]
Comment (susanxhuynh) [reply universe/marathon.json.mustache]: Ok

**Comment by susanxhuynh** [comment]
LGTM. Would you mind editing the title or description to make it clear that UCR is off by default?

File: universe/marathon.json.mustache (hunk 1 above)
Comment (skonto) [reply universe/marathon.json.mustache]: What is the difference?

**Comment by skonto** [comment]
What are the pros and cons for UCR? Customers should pick which?

**Comment by skonto** [comment]
When you pick UCR on the DC/OS UI it complains that docker parameters should be empty.
By default you get:
{
          "key": "user",
          "value": "nobody"
}
It is not so clear.

**Comment by skonto** [comment]
@ArtRand I tried this and it works managed to launch spark shell from the dispatcher container ;)
